src/main/java/com/mongodb/hibernate/internal/type/MongoStructJdbcType.java

```diff
@@ -132,26 +152,52 @@ public BsonDocument createJdbcValue(Object domainValue, WrapperOptions options)
         return result;
     }
 
+    /**
+     * @return The {@linkplain Struct#getAttributes() struct attributes}. Though, the way we support {@link Struct} in
+     *     {@link MongoStructJdbcType} does not involve Hibernate ORM ever {@linkplain Connection#createStruct(String,
+     *     Object[]) creating} one. If we extended {@link org.hibernate.dialect.StructJdbcType}, this could have been
+     *     different.
+     */
     @Override
-    public Object[] extractJdbcValues(Object rawJdbcValue, WrapperOptions options) throws SQLException {
-        if (!(rawJdbcValue instanceof BsonDocument bsonDocument)) {
-            throw fail();
+    public Object @Nullable [] extractJdbcValues(@Nullable Object rawJdbcValue, WrapperOptions options)
+            throws SQLException {
+        if (isNull(rawJdbcValue)) {
+            throw new FeatureNotSupportedException(
+                    "TODO-HIBERNATE-48 https://jira.mongodb.org/browse/HIBERNATE-48 return null");
         }
-        var result = new Object[bsonDocument.size()];
-        var elementIdx = 0;
-        for (var value : bsonDocument.values()) {
-            assertNotNull(value);
-            result[elementIdx++] =
-                    value instanceof BsonDocument ? extractJdbcValues(value, options) : toDomainValue(value);
+        var bsonDocument = assertInstanceOf(rawJdbcValue, BsonDocument.class);
+        var embeddableMappingType = getEmbeddableMappingType();
+        var jdbcValueCount = embeddableMappingType.getJdbcValueCount();
+        var result = new Object[jdbcValueCount];
+        for (int columnIndex = 0; columnIndex < jdbcValueCount; columnIndex++) {
```

Comment:
there are still cases that `var` is not used consistently, though I am perfectly fine with it.